### PR TITLE
feat(gatekeepers): implement a Zupass gatekeeper

### PR DESCRIPTION
**Pull request by gasperpre**
# Description

Implement a MACI gatekeeper which allows singups only to users which have a certain Zupass event ticket.
The gatekeeper was originally implemented in [voto-tech](https://github.com/Vote-tech/voto-tech/blob/final/packages/hardhat/contracts/maci-contracts/gatekeepers/ZupassGatekeeper.sol).

## Additional Notes

The Zupass event must be specified in the contract's constructor and is immutable.

## Confirmation

- [x] I have read and understand MACI's [contributor guidelines](https://maci.pse.dev/docs/contributing) and [code of conduct](https://maci.pse.dev/docs/contributing/code-of-conduct).
- [x] I have read and understand MACI's [GitHub processes](https://github.com/privacy-scaling-explorations/maci/discussions/847).
- [x] I have read and understand MACI's [testing guide](https://maci.pse.dev/docs/testing).

### CODE REVIEW

File: contracts/contracts/gatekeepers/zupass/ZupassGatekeeper.sol [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import { SignUpGatekeeper } from "../SignUpGatekeeper.sol";
+
+import { Groth16Verifier } from "./Groth16Verifier.sol";
+
+/// @title ZupassGatekeeper
+/// @notice This contract allows to gatekeep MACI signups
+/// by requiring new voters to own a certain Zupass event ticket
+contract ZupassGatekeeper is SignUpGatekeeper, Ownable(msg.sender) {
+  /// @notice the Zupass event UUID converted to bigint
+  uint256 private immutable validEventId;
```
Comment (ctrlc03) [inline contracts/contracts/gatekeepers/zupass/ZupassGatekeeper.sol:15]: can we make these public please?

File: contracts/contracts/gatekeepers/zupass/ZupassGatekeeper.sol [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+    // Watermark is stored at index 37
+    // user address converted to bigint is used as the watermark
+    if (_pubSignals[37] != uint256(uint160(_user))) revert InvalidWatermark();
+
+    // Ticket ID is stored at index 0
+    uint256 ticketId = _pubSignals[0];
+    if (registeredTickets[ticketId]) revert AlreadyRegistered();
+
+    registeredTickets[ticketId] = true;
```
Comment (ctrlc03) [inline contracts/contracts/gatekeepers/zupass/ZupassGatekeeper.sol:85]: state changes should happen before external calls are made (so these 3 lines can be moved before proof verification)

File: contracts/contracts/gatekeepers/zupass/Groth16Verifier.sol [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+    or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public
+    License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with snarkJS. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+pragma solidity ^0.8.20;
+
+contract Groth16Verifier {
```
Comment (ctrlc03) [inline contracts/contracts/gatekeepers/zupass/Groth16Verifier.sol:23]: is this verifier contract with vkey on GitHub somewhere or did you generate it?

**Review by ctrlc03 — CHANGES_REQUESTED** [review]
Thanks for this PR, left some comments

File: contracts/contracts/gatekeepers/zupass/Groth16Verifier.sol (hunk 3 above)
Comment (gasperpre) [reply contracts/contracts/gatekeepers/zupass/Groth16Verifier.sol:23]: I think I got it from [zupass-faucet](https://github.com/crisgarner/zupass-faucet/blob/main/packages/hardhat/contracts/Groth16Verifier.sol)

Comment (crisgarner) [reply contracts/contracts/gatekeepers/zupass/Groth16Verifier.sol:23]: This is forked from buidl guild groth16verifier which is using the production zkeys https://github.com/BuidlGuidl/zupass-scaffold-eth-2/blob/main/packages/hardhat/contracts/Groth16Verifier.sol

**Review by ctrlc03 — APPROVED** [review]
Thank you!

**Review by crisgarner — APPROVED** [review]

**Comment by crisgarner** [comment]
Thanks for the PR!

**Pull request merged**